### PR TITLE
DEP: deprecate special.sinc

### LINKED ISSUE #17246: DEP: deprecate numpy aliases scipy.special.sinc
https://github.com/scipy/scipy/issues/16759#issuecomment-1280987784

> FYI: `scipy.special.sinc` is an alias for `numpy.sinc`, so we can ignore that docstring. It would be nice to deprecate and remove `sinc` from the `scipy.special` namespace, but that's a separate issue.
> 
> Similarly, `scipy.integrate.trapezoid` is an alias for `numpy.trapz`.
> 
> I have updated the script to skip these two functions.

### PR DESCRIPTION
**Pull request by dschmitz89**
#### Reference issue
Closes https://github.com/scipy/scipy/issues/17246

#### What does this implement/fix?
`special.sinc` is only an alias for `numpy.sinc`. It is deprecated here, pointing users to use numpy instead.

I did not know how to add the deprecation warning without the `sinc(x)=np.sinc(x)` dance including the copy of the docstring, thats why the diff looks bigger than expected.

### CODE REVIEW

File: scipy/special/_basic.py [hunk 1]
```diff
@@ -107,6 +108,73 @@ def _nonneg_int_or_fail(n, var_name, strict=True):
...
+    >>> plt.title("Sinc Function")
+    >>> plt.ylabel("Amplitude")
+    >>> plt.xlabel("X")
+    >>> plt.show()
+    """
+    _depr_msg = ("scipy.special.sinc is deprecated in SciPy 1.12 and will be "
+                 "removed in SciPy 1.14. We recommend using numpy.sinc "
+                 "instead.")
+    warnings.warn(_depr_msg, DeprecationWarning, stacklevel=2)
+    return np_sinc(x)
```
Comment (j-bowhay) [inline scipy/special/_basic.py:175]:
```suggestion
    return np.sinc(x)
```

File: scipy/special/_basic.py [hunk 2]
```diff
@@ -10,7 +10,8 @@
 from heapq import heapify, heappop
 from numpy import (pi, asarray, floor, isscalar, iscomplex, real,
                    imag, sqrt, where, mgrid, sin, place, issubdtype,
-                   extract, inexact, nan, zeros, sinc)
+                   extract, inexact, nan, zeros)
+from numpy import sinc as np_sinc
```
Comment (j-bowhay) [inline scipy/special/_basic.py:14]:
```suggestion
```

**Review by j-bowhay — COMMENTED** [review]
Thanks @dschmitz89, could you post to the mailing list about this, I doubt there will be any objections given the fact this is just directly wraps numpy but it is good to follow policy regardless

**Comment by Carreau** [comment]
I think you can even say in the deprecation warning that is _was_ and is an alias for numpy.sinc, that may help convince some folks that it is the same instead of wondering what the difference would be.

File: scipy/special/_basic.py [hunk 3]
```diff
@@ -107,6 +107,75 @@ def _nonneg_int_or_fail(n, var_name, strict=True):
     return n
 
 
+def sinc(x):
+    r"""
+    .. deprecated:: 1.12.0
+        scipy.special.sinc is deprecated in SciPy 1.12 and will be removed
+        in SciPy 1.14. As scipy.special.sinc is an alias of numpy.sinc,
+        we encourage using using numpy.sinc instead which returns the
+        same result.
```
Comment (steppi) [inline scipy/special/_basic.py:116]: I think this would be a more standard way to write this.

```suggestion
        `scipy.special.sinc` is a deprecated alias for `numpy.sinc` and will be
        removed in SciPy 1.14. To silence this warning, use `numpy.sinc` instead.
        Deprecated in SciPy 1.12.
```

File: scipy/special/_basic.py [hunk 4]
```diff
@@ -107,6 +107,75 @@ def _nonneg_int_or_fail(n, var_name, strict=True):
...
+    >>> x = np.linspace(-4, 4, 41)
+    >>> plt.plot(x, sinc(x))
+    >>> plt.title("Sinc Function")
+    >>> plt.ylabel("Amplitude")
+    >>> plt.xlabel("X")
+    >>> plt.show()
+    """
+    _depr_msg = ("scipy.special.sinc is deprecated in SciPy 1.12 and will be "
+                 "removed in SciPy 1.14. We recommend using numpy.sinc "
+                 "instead.")
```
Comment (steppi) [inline scipy/special/_basic.py:174]: See above.

**Review by steppi — APPROVED** [review]
Looks good to me. Thanks @dschmitz89!

**Review by j-bowhay — APPROVED** [review]
Thanks @dschmitz89, will leave a couple more days incase of comments from the mailing list

**Comment by rgommers** [comment]
I'm actually having second thoughts about this one. `sinc` is a kinda important special function; PyTorch has added it to its `special` namespace, and it's included in the tentative list of functions for the special functions extension that @mdhaber and @steppi are working on. If it's included there, we really should have it in `scipy.special`. We also have a couple of SciPy-internal usages in `signal`, which will need `special.sinc`'s array API support.

tl;dr I don't think there's an upside of deprecating `sinc` now after so many years, and there are a number of potential downsides.

**Comment by steppi** [comment]
Ah, right. That’s a good point @rgommers. We should only deprecate this if `sinc` is put in the array API proper. I’d be supportive of that but I’m not sure what it would involve.

**Comment by steppi** [comment]
Closing. Our plan is for  `special.sinc`  to dispatch to the correct array API backend based on the input array type, so while it's currently is just an alias to `numpy.sinc`, it won't remain so. Sorry for not paying attention when reviewing earlier.
